lib/rack/session/dalli.rb

```diff
@@ -4,75 +4,143 @@
+      # behavior of Rack::Session, Dalli::Client, and an optional
+      # ConnectionPool. First and foremost, if you wish to instantiate your own
+      # Dalli::Client (or ConnectionPool) and use that instead of letting
+      # Rack::Session::Dalli instantiate it on your behalf, simply pass it in
+      # as the `:cache` option. Please note that you will be responsible for
+      # setting the namespace and any other options on Dalli::Client.
+      #
+      # Secondly, if you're not using the `:cache` option, Rack::Session::Dalli
+      # accepts the same options as Dalli::Client, so it's worth reviewing its
+      # documentation. Perhaps most importantly, if you don't specify a
+      # `:namespace` option, Rack::Session::Dalli will default to using
+      # "rack:session".
+      #
+      # Whether you are using the `:cache` option or not, it is not recommend
+      # to set `:expires_in`. Instead, use `:expire_after`, which will control
+      # both the expiration of the client cookie as well as the expiration of
+      # the corresponding entry in memcached.
+      #
+      # Rack::Session::Dalli also accepts a host of options that control how
+      # the sessions and session cookies are managed, including the
+      # aforementioned `:expire_after` option. Please see the documentation for
+      # Rack::Session::Abstract::Persisted for a detailed explanation of these
+      # options and their default values.
+      #
+      # Finally, if your web application is multithreaded, the
+      # Rack::Session::Dalli middleware can become a source of contention. You
+      # can use a connection pool of Dalli clients by passing in the
+      # `:pool_size` and/or `:pool_timeout` options. For example:
+      #
+      #   use Rack::Session::Dalli,
+      #     :memcache_server => "mc.example.com:1234",
+      #     :pool_size => 10
+      #
+      # You must include the `connection_pool` gem in your project if you wish
+      # to use pool support. Please see the documentation for ConnectionPool
+      # for more information about it and its default options (which would only
+      # be applicable if you supplied one of the two options, but not both).
+      #
       def initialize(app, options={})
         super
-        @mutex = Mutex.new
-        mserv = @default_options[:memcache_server]
-        mopts = @default_options.reject{|k,v| !DEFAULT_OPTIONS.include? k }
-        @pool = options[:cache] || ::Dalli::Client.new(mserv, mopts)
-        @pool.alive!
-      end
 
-      if defined?(Abstract::Persisted)
-        def find_session(req, sid)
-          get_session req.env, sid
-        end
+        # determine the default TTL for newly-created sessions
+        @default_ttl = ttl @default_options[:expire_after]
 
-        def write_session(req, sid, session, options)
-          set_session req.env, sid, session, options
-        end
+        mopts = DEFAULT_DALLI_OPTIONS.merge \
+          options.reject {|k, _| DEFAULT_OPTIONS.key? k }
```

Comment:
I had to fundamentally altered the options parsing logic of Rack::Session::Dalli. The original version leverages Rack::Session::Abstract::Persisted's setup to handle the options for both Rack::Session::Abstract::Persisted and Dalli::Client, using `DEFAULT_OPTIONS` as a whitelist for both. Any options not added to the whitelist, such as `:failover` or `:serializer`, will not be passed to Dalli::Client. This severely limits the ability of the user to customize the Dalli::Client instance being used by Rack::Session::Dalli; only `:namespace` and the Memcached server can be specified! It also prevents us from passing `:pool_size` and `:pool_timeout` options through to the ConnectionPool setup.

One solution would have been to add the Dalli::Client and ConnectionPool options that we're "okay" with to `DEFAULT_OPTIONS`. I started doing this and after reviewing the 19 additional options available to Dalli::Client and adding them to the list, plus the two for ConnectionPool, it occurred to me that this is kind of silly. In my mind, any options the user passes to Rack::Session::Dalli at `use` time should pass to Rack::Session::Abstract::Persisted first, and any leftover options should pass to Dalli::Client, with each class handling its own setup. We can apply any defaults specific to Rack::Session::Dalli in our logic (such as the default namespace), and peel off the Memcached server name(s) and any ConnectionPool options for "special" handling.

That's what the new options parsing does:

1. Call `super`. Rack::Session::Abstract::Persisted will use `DEFAULT_OPTIONS` to build `@default_options`. These are options that primarily control how the session cookie is generated.
1. Remove any elements from the options hash that are/were handled by Rack::Session::Abstract::Persisted, leaving only the Memcached server and options for Dalli::Client and ConnectionPool.
1. Apply the defaults specified in `DEFAULT_DALLI_OPTIONS` to the remaining options.
1. Remove the Memcached server name(s) and ConnectionPool options (as these shouldn't be passed to Dalli::Client in the options hash). 

So that's why we're rejecting instead of selecting options based on `DEFAULT_OPTIONS`. It's no longer a whitelist; it's now a way of differentiating between options intended for Rack::Session::Abstract::Persisted and options intended for Dalli::Client and ConnectionPool.